File: backend/app/api/v1/projects.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_active_user, get_db, get_project_member
from app.models.user import User
from app.schemas.project import (
    ProjectCreate,
    ProjectListResponse,
    ProjectResponse,
    ProjectUpdate,
)
from app.services.project_service import ProjectService

router = APIRouter()


@router.post(
    "",
    response_model=ProjectResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_project(
    data: ProjectCreate,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    service = ProjectService(db)
    return await service.create_project(data, current_user)


@router.get("", response_model=ProjectListResponse)
async def list_projects(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    service = ProjectService(db)
    projects = await service.list_user_projects(current_user.id)
    return ProjectListResponse(items=projects, total=len(projects))


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    service = ProjectService(db)
    project = await service.get_project(project_id)
    if project is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Project not found"
        )
    return ProjectResponse.model_validate(project)


@router.patch("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: uuid.UUID,
    data: ProjectUpdate,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    # Verify membership
    await get_project_member(project_id, current_user, db)
    service = ProjectService(db)
    try:
        return await service.update_project(project_id, data)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e)
        )


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def archive_project(
    project_id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    await get_project_member(project_id, current_user, db)
    service = ProjectService(db)
    try:
        await service.archive_project(project_id)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e)
        )

File: frontend/src/api/projects.ts
import apiClient from './client'

export interface ProjectCreate {
  name: string
  description?: string
  country: string
  admin_level?: number
  year: number
}

export interface ProjectResponse {
  id: string
  name: string
  description: string | null
  country: string
  admin_level: number
  year: number
  status: string
  created_at: string
  updated_at: string
}

export interface ProjectListResponse {
  items: ProjectResponse[]
  total: number
}

export const projectsApi = {
  list() {
    return apiClient.get<ProjectListResponse>('/projects')
  },

  create(data: ProjectCreate) {
    return apiClient.post<ProjectResponse>('/projects', data)
  },

  get(projectId: string) {
    return apiClient.get<ProjectResponse>(`/projects/${projectId}`)
  },

  update(projectId: string, data: Partial<ProjectCreate>) {
    return apiClient.patch<ProjectResponse>(`/projects/${projectId}`, data)
  },

  archive(projectId: string) {
    return apiClient.delete(`/projects/${projectId}`)
  },
}
